File: create_svg.py
"""Generate OLL diagrams as SVG files.

The input format is a dictionary with five keys:

    U: nine stickers of the upper face, read left-to-right, top-to-bottom
    B: three stickers behind the upper face (drawn above U)
    F: three stickers in front of the lower face (drawn below U)
    L: three stickers on the left side, read top-to-bottom
    R: three stickers on the right side, read top-to-bottom

Use ``Y`` for a yellow sticker and ``N`` for a non-yellow sticker.
The module has no third-party dependencies.
"""

from __future__ import annotations

from pathlib import Path
from typing import Mapping, Sequence
from xml.etree.ElementTree import Element, SubElement, tostring
from xml.dom import minidom


COLOR_MAP = {
    "Y": "#FFE94A",  # yellow sticker
    "N": "#D9D9D9",  # non-yellow sticker
}

# Canvas and layout. Values are chosen to match the proportions in the
# reference image: large rounded central stickers and narrow side stickers.
SIZE = 512
CANVAS_COLOR = "#1E1E1E"
BACKGROUND = "#111111"
STICKER_GAP = 7
CELL = 64
STICKER_RADIUS = 12
EDGE_STICKER = 18
EDGE_RADIUS = 9

FACE_X = 133
FACE_Y = 133
FACE_SIZE = 3 * CELL + 2 * STICKER_GAP  # 206


def _rect(parent: Element, x: int, y: int, width: int, height: int,
          fill: str, radius: int) -> None:
    SubElement(
        parent,
        "rect",
        {
            "x": str(x),
            "y": str(y),
            "width": str(width),
            "height": str(height),
            "rx": str(radius),
            "ry": str(radius),
            "fill": fill,
        },
    )


def _validate(state: Mapping[str, Sequence[str]]) -> None:
    expected_lengths = {"U": 9, "B": 3, "F": 3, "L": 3, "R": 3}
    missing = set(expected_lengths) - set(state)
    extra = set(state) - set(expected_lengths)
    if missing or extra:
        details = []
        if missing:
            details.append(f"missing keys: {sorted(missing)}")
        if extra:
            details.append(f"unknown keys: {sorted(extra)}")
        raise ValueError("Invalid OLL dictionary (" + "; ".join(details) + ")")

    for face, length in expected_lengths.items():
        stickers = state[face]
        if len(stickers) != length:
            raise ValueError(f"{face} must contain {length} values, got {len(stickers)}")
        invalid = set(stickers) - set(COLOR_MAP)
        if invalid:
            raise ValueError(f"{face} contains invalid values: {sorted(invalid)}; use Y or N")


def generate_svg(state: Mapping[str, Sequence[str]], filename: str | Path) -> Path:
    """Render one OLL dictionary and save it to *filename*.

    ``B`` is rendered above the U face and ``F`` below it, as in the
    convention described in the question.
    """
    _validate(state)
    filename = Path(filename)

    svg = Element(
        "svg",
        {
            "xmlns": "http://www.w3.org/2000/svg",
            "width": str(SIZE),
            "height": str(SIZE),
            "viewBox": f"0 0 {SIZE} {SIZE}",
        },
    )
    # Dark canvas plus the compact black frame behind the cube net.
    # _rect(svg, 0, 0, SIZE, SIZE, CANVAS_COLOR, 0)
    # _rect(
    #     svg,
    #     FACE_X - EDGE_STICKER - 2 * STICKER_GAP,
    #     FACE_Y - EDGE_STICKER - 2 * STICKER_GAP,
    #     FACE_SIZE + 2 * (EDGE_STICKER + 2 * STICKER_GAP),
    #     FACE_SIZE + 2 * (EDGE_STICKER + 2 * STICKER_GAP),
    #     BACKGROUND,
    #     16,
    # )

    columns = [FACE_X + i * (CELL + STICKER_GAP) for i in range(3)]
    rows = [FACE_Y + i * (CELL + STICKER_GAP) for i in range(3)]

    # U: the 3x3 face.
    for index, value in enumerate(state["U"]):
        row, column = divmod(index, 3)
        _rect(svg, columns[column], rows[row], CELL, CELL,
              COLOR_MAP[value], STICKER_RADIUS)

    # B is behind U and therefore goes at the top; F goes at the bottom.
    top_y = FACE_Y - STICKER_GAP - EDGE_STICKER
    bottom_y = FACE_Y + FACE_SIZE + STICKER_GAP
    for column, value in enumerate(state["B"]):
        _rect(svg, columns[column], top_y, CELL, EDGE_STICKER,
              COLOR_MAP[value], EDGE_RADIUS)
    for column, value in enumerate(state["F"]):
        _rect(svg, columns[column], bottom_y, CELL, EDGE_STICKER,
              COLOR_MAP[value], EDGE_RADIUS)

    # L and R: the side strips, from top to bottom.
    left_x = FACE_X - STICKER_GAP - EDGE_STICKER
    right_x = FACE_X + FACE_SIZE + STICKER_GAP
    for row, value in enumerate(state["L"]):
        _rect(svg, left_x, rows[row], EDGE_STICKER, CELL,
              COLOR_MAP[value], EDGE_RADIUS)
    for row, value in enumerate(state["R"]):
        _rect(svg, right_x, rows[row], EDGE_STICKER, CELL,
              COLOR_MAP[value], EDGE_RADIUS)

    filename.parent.mkdir(parents=True, exist_ok=True)
    xml = minidom.parseString(tostring(svg, encoding="utf-8")).toprettyxml(
        indent="  ", encoding="utf-8"
    )
    filename.write_bytes(xml)
    return filename


def generate_many(olls: Mapping[str, Mapping[str, Sequence[str]]],
                  output_dir: str | Path = "svg") -> list[Path]:
    """Generate one ``<name>.svg`` file for every named OLL dictionary."""
    output_dir = Path(output_dir)
    return [
        generate_svg(state, output_dir / f"{name}.svg")
        for name, state in olls.items()
    ]


# Example from the question. Add more dictionaries to OLLS to generate them
# all with one run of the script.
oll_51 = {
    "U": ["N", "Y", "N", "N", "Y", "N", "N", "Y", "N"],
    "F": ["Y", "N", "N"],
    "R": ["Y", "Y", "Y"],
    "B": ["Y", "N", "N"],
    "L": ["N", "Y", "N"],
}

OLLS = {"oll_51": oll_51}


if __name__ == "__main__":
    created = generate_many(OLLS)
    for path in created:
        print(f"Created: {path}")

File: create_svg.js
/**
 * Generate OLL situations as SVG diagrams.
 *
 * A situation is described by five arrays:
 *
 *   U - the 3x3 upper face, left-to-right and top-to-bottom
 *   B - the three stickers behind the upper face (shown above U)
 *   F - the three stickers in front of the upper face (shown below U)
 *   L - the three stickers on the left, top-to-bottom
 *   R - the three stickers on the right, top-to-bottom
 *
 * Y means a yellow sticker and N means a non-yellow sticker.
 *
 * Run:
 *   node create_svg.js
 *
 * The example situation is written to svg/oll_01.svg. More situations
 * can be added to OLLS and will be generated in the same run.
 */

const fs = require("fs");
const path = require("path");

const COLORS = Object.freeze({
  Y: "#FFFF00",
  N: "#8D8D8D",
  BACKGROUND: "#111111",
  FRAME: "#000000",
});

// These values match the proportions of the reference diagram.
const SVG_WIDTH = 637;
const SVG_HEIGHT = 563;
const CELL = 123;
const EDGE_STICKER = 45;
const GAP = 13;
const FACE_X = 106;
const FACE_Y = 86;
const FACE_SIZE = CELL * 3 + GAP * 2;
const FACE_RADIUS = 16;
const EDGE_RADIUS = 8;

function escapeXml(value) {
  return String(value)
    .replaceAll("&", "&amp;")
    .replaceAll('"', "&quot;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;");
}

function rect(x, y, width, height, fill, radius = 0) {
  return `<rect x="${x}" y="${y}" width="${width}" height="${height}"` +
    ` rx="${radius}" ry="${radius}" fill="${escapeXml(fill)}"/>`;
}

function validate(state) {
  const expectedLengths = { U: 9, B: 3, F: 3, L: 3, R: 3 };
  const keys = Object.keys(state ?? {});
  const missing = Object.keys(expectedLengths).filter((key) => !(key in (state ?? {})));
  const extra = keys.filter((key) => !(key in expectedLengths));

  if (missing.length || extra.length) {
    const details = [];
    if (missing.length) details.push(`missing keys: ${missing.join(", ")}`);
    if (extra.length) details.push(`unknown keys: ${extra.join(", ")}`);
    throw new Error(`Invalid OLL dictionary (${details.join("; ")})`);
  }

  for (const [face, expectedLength] of Object.entries(expectedLengths)) {
    const stickers = state[face];
    if (!Array.isArray(stickers) || stickers.length !== expectedLength) {
      throw new Error(`${face} must contain exactly ${expectedLength} values`);
    }

    const invalid = stickers.filter((value) => value !== "Y" && value !== "N");
    if (invalid.length) {
      throw new Error(
        `${face} contains invalid values: ${[...new Set(invalid)].join(", ")}; use Y or N`,
      );
    }
  }
}

function generateSvg(state, filename) {
  validate(state);

  const columns = [0, 1, 2].map((index) => FACE_X + index * (CELL + GAP));
  const rows = [0, 1, 2].map((index) => FACE_Y + index * (CELL + GAP));
  const elements = [
    // Background and the black cross-shaped frame behind the stickers.
    rect(99, 21, 409, 59, COLORS.FRAME),
    rect(41, 79, 525, 409, COLORS.FRAME),
    rect(99, 488, 409, 59, COLORS.FRAME),
  ];

  // U: the 3x3 upper face.
  state.U.forEach((value, index) => {
    const row = Math.floor(index / 3);
    const column = index % 3;
    elements.push(
      rect(columns[column], rows[row], CELL, CELL, COLORS[value], FACE_RADIUS),
    );
  });

  // B is above U and F is below U.
  const topY = FACE_Y - GAP - EDGE_STICKER;
  const bottomY = FACE_Y + FACE_SIZE + GAP;
  state.B.forEach((value, column) => {
    elements.push(
      rect(columns[column], topY, CELL, EDGE_STICKER, COLORS[value], EDGE_RADIUS),
    );
  });
  state.F.forEach((value, column) => {
    elements.push(
      rect(columns[column], bottomY, CELL, EDGE_STICKER, COLORS[value], EDGE_RADIUS),
    );
  });

  // L and R are vertical strips, read from top to bottom.
  const leftX = FACE_X - GAP - EDGE_STICKER;
  const rightX = FACE_X + FACE_SIZE + GAP;
  state.L.forEach((value, row) => {
    elements.push(
      rect(leftX, rows[row], EDGE_STICKER, CELL, COLORS[value], EDGE_RADIUS),
    );
  });
  state.R.forEach((value, row) => {
    elements.push(
      rect(rightX, rows[row], EDGE_STICKER, CELL, COLORS[value], EDGE_RADIUS),
    );
  });

  const svg = [
    '<?xml version="1.0" encoding="utf-8"?>',
    `<svg xmlns="http://www.w3.org/2000/svg" width="${SVG_WIDTH}" height="${SVG_HEIGHT}" viewBox="0 0 ${SVG_WIDTH} ${SVG_HEIGHT}">`,
    ...elements.map((element) => `  ${element}`),
    "</svg>",
    "",
  ].join("\n");

  const outputPath = path.resolve(filename);
  fs.mkdirSync(path.dirname(outputPath), { recursive: true });
  fs.writeFileSync(outputPath, svg, "utf8");
  return outputPath;
}

function generateMany(olls, outputDir = "svg") {
  return Object.entries(olls).map(([name, state]) =>
    generateSvg(state, path.join(outputDir, `${name}.svg`)),
  );
}

const OLL_01 = {
    "U": [
      "Y",
      "N",
      "Y",
      "N",
      "Y",
      "Y",
      "N",
      "Y",
      "N"
    ],
    "F": [
      "Y",
      "N",
      "Y"
    ],
    "R": [
      "N",
      "N",
      "N"
    ],
    "B": [
      "N",
      "Y",
      "N"
    ],
    "L": [
      "N",
      "Y",
      "N"
    ]
  };

const OLLS = { oll_01: OLL_01 };

if (require.main === module) {
  for (const filename of generateMany(OLLS)) {
    console.log(`Created: ${filename}`);
  }
}

module.exports = {
  COLORS,
  OLL_01,
  OLLS,
  generateSvg,
  generateMany,
  validate,
};
